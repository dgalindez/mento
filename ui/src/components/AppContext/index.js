import { createContext, useEffect } from 'react';

import useGeolocation from '../../hooks/useGeolocation';
import useLocale from '../../hooks/useLocale';
import useWeatherApi from '../../hooks/useWeatherApi';

const AppContext = createContext({
  geolocation: {},
  locale: {},
  weather: {},
});

export const ContextComponent = ({ children }) => {
  const geolocation = useGeolocation();
  const locale = useLocale();
  const weather = useWeatherApi();

  useEffect(() => {
    if (!geolocation?.failed) {
      weather?.getLocationData(geolocation.lat, geolocation.lon);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [geolocation.failed, geolocation.lat, geolocation.lon]);

  return (
    <AppContext.Provider value={{ geolocation, locale, weather }}>
      {children}
    </AppContext.Provider>
  );
};

export default AppContext;
