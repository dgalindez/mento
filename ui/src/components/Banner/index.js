import { useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';

import AppContext from '../AppContext';

const Banner = () => {
  const { t } = useTranslation();
  const {
    geolocation: {
      failed,
      loading,
    } = {},
    weather: {
      error,
      loading: weatherLoading,
    },
  } = useContext(AppContext);
  const [dismissed, setDismissed] = useState(false);
  const [dismissedError, setDismissedError] = useState(false);

  const dismiss = () => setDismissed(true);
  const dismissError = () => setDismissedError(true);

  return (
    <>
      {!loading && failed && !dismissed && (
        <article className="message is-warning">
          <div className="message-header">
            <p>{t('banner.loading.failed.title')}</p>
            <button className="delete" aria-label="delete" onClick={dismiss}>
            </button>
          </div>
          <div className="message-body">
            {t('banner.loading.failed.body')}
          </div>
        </article>
      )}
      {!weatherLoading && error && !dismissedError && (
        <article className="message is-error">
          <div className="message-header">
            <p>{t(`${error}`)}</p>
            <button className="delete" aria-label="delete" onClick={dismissError}>
            </button>
          </div>
        </article>
      )}
    </>
  );
};

export default Banner;
