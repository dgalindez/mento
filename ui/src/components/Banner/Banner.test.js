import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../utils/i18nTestHelper';
import AppContext from '../AppContext';
import Banner from '.';

describe('Banner', () => {
  let contextData = {};

  beforeEach(() => {
    contextData = {
      geolocation: {
        loading: false,
        failed: false,
      },
      weather: {
        error: '',
        loading: false,
      },
    }
  });

  it('renders nothing with empty data', () => {
    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.queryByRole('article')).toBeNull();
  });

  it('renders nothing if loading', () => {
    contextData.geolocation.loading = true;

    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.queryByRole('article')).toBeNull();
  });

  it('renders nothing if weather loading', () => {
    contextData.weather.loading = true;

    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.queryByRole('article')).toBeNull();
  });

  it('renders if failed', () => {
    contextData.geolocation.failed = true;

    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.getByRole('article')).toMatchSnapshot();
  });

  it('renders if weather is not loading and has error', () => {
    contextData.weather.loading = false;
    contextData.weather.error = 'error';

    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(screen.getByRole('article')).toMatchSnapshot();
  });

  it('renders nothing if dismissed', async () => {
    contextData.geolocation.failed = true;

    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    userEvent.click(screen.getByRole('button'))
    await waitFor(() => expect(screen.queryByRole('article')).toBeNull());
  });

  it('renders nothing if dismissed error', async () => {
    contextData.weather.loading = false;
    contextData.weather.error = 'error';

    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <Banner />
        </AppContext.Provider>
      </I18nextProvider>
    );

    userEvent.click(screen.getByRole('button'))
    await waitFor(() => expect(screen.queryByRole('article')).toBeNull());
  });
});
