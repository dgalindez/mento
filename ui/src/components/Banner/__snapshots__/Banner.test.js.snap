// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner renders if failed 1`] = `
<article
  class="message is-warning"
>
  <div
    class="message-header"
  >
    <p>
      banner.loading.failed.title
    </p>
    <button
      aria-label="delete"
      class="delete"
    />
  </div>
  <div
    class="message-body"
  >
    banner.loading.failed.body
  </div>
</article>
`;

exports[`Banner renders if weather is not loading and has error 1`] = `
<article
  class="message is-error"
>
  <div
    class="message-header"
  >
    <p>
      error
    </p>
    <button
      aria-label="delete"
      class="delete"
    />
  </div>
</article>
`;
